File: PythonCodes/Chapter2/time_complexity.py
def constant(n: int) -> int:
    """常数阶"""
    count = 0
    size = 100000
    for _ in range(size):
        count += 1
    
    return count

def linear(n: int) -> int:
    """线性阶"""
    count = 0
    for _ in range(n):
        count += 1
    return count

def quadratic(n: int) -> int:
    """平方阶"""
    count = 0 # 计数器
    for i in range(n):
        for j in range(n):
            count += 1
    return count

def bubble_sort(nums: list[int]) -> int:
    """平方阶（冒泡排序）"""
    count = 0 # 计数器
    for i in range(len(nums) - 1, 0, -1):
        for j in range(i):
            if nums[j] > nums[j + 1]:
                temp: int = nums[j]
                nums[j] = nums[j+1]
                nums[j + 1] = temp
                count += 3
    
    return count
             
def exponential(n: int) -> int:
    """指数阶（循环实现）"""
    count = 0
    base = 1
    for _ in range(n):
        for _ in range(base):
            count += 1
        base *= 2
    
    return count   

def exp_recur(n: int) -> int:
    """指数阶（递归实现）"""
    if n == 1:
        return 1
    return exp_recur(n - 1) + exp_recur(n - 1) + 1

def logarithmic(n: int) -> int:
    """对数阶（循环实现）"""
    count = 0
    while n > 1:
        n = n / 2
        count += 1
    
    return count

"""Driver Code"""
if __name__ == "__main__":
    n = 3
    print("输入数据大小 n = ", n)
    
    count = logarithmic(n)
    print("常数阶的操作数量 = ", count)
    
    

File: CppCodes/Chapter2/time_complexity.cpp
#include <iostream>
#include <vector>

/* 常数阶 */
int constant(int n)
{
    int count = 0;
    int size = 100000;
    for (int i = 0; i < size; i++)
    {
        count++;
    }

    return count;
}

/* 线性阶 */
int linear(int n) // 输入数据大小为 n
{
    int count = 0;
    for (int i = 0; i < n; i++)
    {
        count++;
    }
    return count;
}

/* 线性阶（遍历数组） */
int arrayTraversal(std::vector<int> &nums)
{
    int count = 0;
    for (int num : nums) // 输入数据大小为数组长度
    {
        count++;
    }
    return count;
}

/* 平方阶 */
int quadratic(int n)
{
    int count = 0;
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            count++;
        }
    }
    return count;
}

/* 平方阶（冒泡排序） */
int bubbleSort(std::vector<int> &nums)
{
    int count = 0; // 计数器
    // 外循环：未排序区间为 [0, i]
    for (int i = nums.size() - 1; i > 0; i--)
    {
        // 内循环：将未排序区间 [0, i] 中的最大元素交换至该区间的最右端
        for (int j = 0; j < i; j++)
        {
            if (nums[j] > nums[j + 1])
            {
                int temp = nums[j];
                nums[j] = nums[j + 1];
                nums[j + 1] = temp;
                count += 3;
            }
        }
    }
    return count;
}

/* 指数阶（循环实现）*/
int exponential(int n)
{
    int count = 0, base = 1;
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < base; j++)
        {
            count++;
        }
        base *= 2; // 每一层的个数都乘以 2
    }
    return count;
}

/* 指数阶（递归实现） */
int expRecur(int n)
{
    if (n == 1)
        return 1;
    return expRecur(n - 1) + expRecur(n - 1) + 1;
}

/* 对数阶（循环实现）*/
int logarighmic(int n)
{
    int count = 0;
    while (n > 1)
    {
        n = n / 2;
        count++;
    }
    return count;
}

int main()
{
    int count = exponential(8);
    std::cout << count << std::endl;
}
